PPTP 1636 Nominated partners must validate their email addresses

tonymccrae-ee opened the pull request:
### Description of Work carried through

![Screenshot from 2022-02-23 16-48-17](https://user-images.githubusercontent.com/95688374/155366738-7a4a7e8d-cfb2-41e9-b4c7-79e84d9112fa.png)
![Screenshot from 2022-02-23 16-33-34](https://user-images.githubusercontent.com/95688374/155366739-f5c8d081-0009-45b0-9f19-5ac67fa31e7c.png)


Starting with `AmendEmailAddressController.scala` as the best pre existing example of an email verification flow, 
we have stripped it for parts and collected the common actions in `EmailVerificationActions`.

As partner email verification applies to registrations and amendments we need away of persisting email verification journey data which will work from registration and amendment journeys.

The lowest common denominator seems to be `RegistrationUpdater`. 
The side effects of using this are that backend registration needs to persist the emali verification prospective email address and journey id on the mongo registration. Amendments are probably making a full EMTP submission when storing the prospective email. Both of these are side effects of choosing to use the Registration so store this temporary data.

The 3 controllers which deal with partner emails extended EmailVerificationActions and provide the email verification routes.
(`AddPartnerContactDetailsEmailAddressController` may not strictly need this) 

Marked 3 TODO items relating to the existing code which we can address is anyone has opinions.

#### Check list 
 - [x] `./precheck` was executed (Integration/Component/Unit tests)
 - [x] `sbt scalafmt test:scalafmt` was executed
 - [ ] Required Environment Config has been amended/added
 - [ ] Links to dependencies have been included (BE/FE work)
 - [ ] User Acceptance Tests (UAT) were run locally and have passed.
 - [ ] No Accessibility errors/warnings reported by Wave

RyanKellyHMRC commented on app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/EmailVerificationActions.scala:
```diff
@@ -0,0 +1,218 @@
+    email: EmailAddress,
+    emailVerificationJourneyId: String
+  )(implicit journeyRequest: JourneyRequest[AnyContent], hc: HeaderCarrier): Future[Registration] =
+    // By updating and persisting the entire registration; all the way back to ETMP in the case of an amendment
```
I'm pretty certain we're good here and that only the local amendment registration is being affected.

tonymccrae-ee commented on test/uk/gov/hmrc/plasticpackagingtax/registration/controllers/amendment/partner/AmendPartnerContactDetailsControllerSpec.scala:
```diff
@@ -456,6 +478,167 @@ class AmendPartnerContactDetailsControllerSpec
+        )
+
+        // Would be nice to check that email verification journey details were set on the registration
+        // but can't work out how to get hold of the updated registration in this context
```
Patch applied; thanks.

tonymccrae-ee commented on app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/contact/ContactDetailsEmailAddressController.scala:
```diff
@@ -147,7 +147,7 @@ class ContactDetailsEmailAddressController @Inject() (
       email,
       credId,
-      "/register-for-plastic-packaging-tax/primary-contact-details"
+      "/register-for-plastic-packaging-tax/primary-contact-details" // TODO this is not a valid route?
```
I will fix this.

tonymccrae-ee commented on conf/app.routes:
```diff
@@ -116,10 +116,20 @@ POST       /partner-job-title                   uk.gov.hmrc.plasticpackagingtax.
+GET        /partner-email-address-confirm-email-code/:partnerId         uk.gov.hmrc.plasticpackagingtax.registration.controllers.partner.PartnerEmailAddressController.confirmExistingPartnerEmailCode(partnerId: String)
+POST       /partner-email-address-confirm-email-code/:partnerId         uk.gov.hmrc.plasticpackagingtax.registration.controllers.partner.PartnerEmailAddressController.checkExistingPartnerEmailVerificationCode(partnerId: String)
+GET        /partner-email-confirmed/:partnerId                          uk.gov.hmrc.plasticpackagingtax.registration.controllers.partner.PartnerEmailAddressController.emailVerifiedExistingPartner(partnerId: String)
+GET        /partner-email-confirm-change/:partnerId                     uk.gov.hmrc.plasticpackagingtax.registration.controllers.partner.PartnerEmailAddressController.confirmEmailUpdateExistingPartner(partnerId: String)
```
The too many attempts page shows generic copy with no journey specific links.
I could have used a project wide one but think someone might call for journey specific link in the future.

Therefore, I've provided a too many attempts route on each on the effected controllers.
In this specific controller I decided 1 is probably close enough.

tonymccrae-ee commented:
Addressed feedback.
Completed of deleted the unrelated TODO I'd marked.

Wrapped confirm update email address endpoints with a check to see if the user really completed email verification.

RyanKellyHMRC reviewed (approved):
LGTM
